### PR TITLE
Fix S2583/S2589 FN and FP: Add support for relational pattern

### LINKED ISSUE #8493: Fix S2583/S2589 FN and FP: Add support for relational pattern
Add support for RelationalPattern in `IsPattern` implementation.

It should fix this FN in S2583/S25889
```
        void ArithmeticComparisonAndPattern()
        {
            int? c = null;
            if (c is > 10 and < 100)    // FN
            {
                DoSomething();
            }
        }
```
and FPs caused by the missing evaluation of NumberConstraint.

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
Fixes #8493

Review per commit shows the evolution

### CODE REVIEW

File: analyzers/its/expected/Net5/Net5--net5.0-S2589.json [hunk 1]
```diff
@@ -4,6 +4,58 @@
...
+"endLine":  32,
+"endColumn":  13
+}
+}
+},
+{
+"id":  "S2589",
+"message":  "Change this condition so that it does not always evaluate to 'False'.",
+"location":  {
+"uri":  "https://github.com/SonarSource/sonar-dotnet/blob/master/analyzers/its/sources/Net5/Net5/Main.cs#L37",
```
Comment (cristian-ambrosini-sonarsource) [inline analyzers/its/expected/Net5/Net5--net5.0-S2589.json:33]: Why are we raising twice for each line? #L37 and #L32

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/ConditionEvaluatesToConstant.CSharp9.cs [hunk 2]
```diff
@@ -31,16 +31,29 @@ void IsNotPattern()
...
+        }
+
+        void ArithmeticComparisonAndPattern_Null()
+        {
+            int? c = null;
+            if (c is > 10 and < 100)    // Noncompliant
+            {
+                DoSomething();          // Secondary
             }
         }
```
Comment (cristian-ambrosini-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/ConditionEvaluatesToConstant.CSharp9.cs:47]: This is a duplicate of the test above

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/IsPattern.cs [hunk 3]
```diff
@@ -240,4 +241,21 @@ BoolConstraint CombineOr()
             }
         }
     }
+
+    private static SymbolicConstraint BoolConstraintFromRelationalPattern(SymbolicValue value, IRelationalPatternOperationWrapper pattern)
```
Comment (cristian-ambrosini-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/IsPattern.cs:245]: Nice!

**Review by cristian-ambrosini-sonarsource — CHANGES_REQUESTED** [review]
Duplicate test, but other than that the logic looks good to my SE newbie's eyes

File: analyzers/its/expected/Net5/Net5--net5.0-S2589.json (hunk 1 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/its/expected/Net5/Net5--net5.0-S2589.json:33]: That's a good question. I need to investigae

Comment (pavel-mikula-sonarsource) [reply analyzers/its/expected/Net5/Net5--net5.0-S2589.json:33]: .NET 5 SDK had a bug. Some new nodes, like CompilationUnit due to top-level statement were called twice.
So we analyze it twice.

This only appears when you have global.json hardcoded .NET 5 SDK, or you have .NET 5 SDK as the latest SDK. It's fine if you have newer SDK and target net5.0.

I'll leave it like this.

**Review by cristian-ambrosini-sonarsource — APPROVED** [review]
LGTM!

**Pull request merged**